encryptionkeyrotation/README.md

```diff
@@ -0,0 +1,95 @@
+# CSI Addons Operation: EncryptionKeyRotation
+
+## Terminology
+
+| Term     | Definition                                                                            |
+| -------- | ------------------------------------------------------------------------------------- |
+| VolumeID | The identifier of the volume generated by the plugin.                                 |
+| CO       | Container Orchestration system that communicates with plugins using CSI service RPCs. |
+| SP       | Storage Provider, the vendor of a CSI plugin implementation.                          |
+| RPC      | [Remote Procedure Call](https://en.wikipedia.org/wiki/Remote_procedure_call).         |
+
+## Objective
+
+Define an extension to the CSI specification that will enable storage vendors
+(SP) to develop controllers/plugins that can rotate the Encryption Keys (KEK)
+of existing volumes that are encrypted.
+
+### Goals in MVP
+
+The new extension will define a procedure that
+
+- can be called for existing encrypted volumes
+- functions for Block mode volumes
+- interacts with the Node-Plugin
+- makes it possible for the SP to rotate encrpytion keys on volumes
+
+### Non-Goals in MVP
+
+- Rotating encrpytion keys for inactive volumes
+
+## Solution Overview
+
+This specification defines an interface along with the minimum operational and
+packaging recommendations for a storage provider (SP) to implement a key rotate operation for volumes. The interface declares the RPCs that a plugin
+MUST expose.
+
+## RPC Interface
+
+- **Node Service**: The Node plugin MAY implement this RPC.
+
+```protobuf
+syntax = "proto3";
+package encryptionkeyrotation;
+
+import "github.com/container-storage-interface/spec/lib/go/csi/csi.proto";
+import "google/protobuf/descriptor.proto";
+
+option go_package = "github.com/csi-addons/spec/lib/go/encryptionkeyrotation";
+
+// EncryptionKeyRotationController holds the RPC method for running
+// key rotation operation on the volume.
+service EncryptionKeyRotationController {
+    // EncryptionKeyRotate is a procedure that is called on the CSI Nodeplugin
+    rpc EncryptionKeyRotate(EncryptionKeyRotateRequest)
+    returns (EncryptionKeyRotateResponse){}
+}
+```
+
+### EncryptionKeyRotateRequest
+
+```protobuf
+// EncryptionKeyRotateRequest contains the information needed to identify
+// the volume by the SP and access any backend services so that the key can be
+// rotated.
+message EncryptionKeyRotateRequest {
```

Comment:
should this not include the current and a new key to be set?

Reply:
Nah, they are generated on the ceph-csi end of things.

Reply:
Maybe for Ceph-CSI, but this specification should be usable by other CSI-drivers as well. Even within Ceph-CSI, users can store the encryption key in a secret, and Ceph-CSI reads that. If a user wants to replace the key in the secret by an other key, at least the new key needs to be provided (and as a security verification the old key would be useful).

Reply:
In most of the cases the keys are handled by a KMS on the csi-driver end. In such cases one can leverage the `params` field maybe?

Reply:
I still think it is required to pass the new key (annotated as `csi.v1.csi_secret`). The original key can probably be obtained by the CSI-driver the way that they do already.

A new key can be made OPTIONAL, if missing the CSI-driver should generate one itself (or have the KMS do that).

Users may have the requirement to pass the new key themselves. This is at least true for Ceph-CSI where a key can be stored in a Kubernetes Secret. The key can not be updated in the Secret before doing the `EncryptionKeyRotate`, as that would prevent retrieving the original key which is needed for authenticating before rotating (I am thinking of the `cryptsetup` command here).

Reply:
As the implementation is entirely up to the SP, can the SP not use the existing secrets map and provide the key there?

Reply:
I do not know, maybe it is possible? Can you give an example workflow how a user would would be able to rotate the key with a chosen value?